### ISSUE TITLE
non-blocking receive method

### ISSUE THREAD

**Issue by Odebe** [crystal-lang/crystal#8222]
Issue for discussing [this PR](https://github.com/crystal-lang/crystal/pull/8218)

I think this is good idea because it allow us to create fibers with loops that handle incoming messages from multiple channels without blocking fiber on each channel.

**Comment by asterite** [comment]
I think in your real example you can use regular `receive`, then replace this:

```crystal
Server.new.start
```

with this:

```crystal
spawn { Server.new.start }

sleep
```

In the other example you have I don't think doing polling (checking once every second) instead of blocking the fiber until it really has work to do is better.

**Comment by Odebe** [comment]
```
spawn { Server.new.start }

sleep
```
I don't understand how it should solve my problem, I want server to process messages from multiple channels in one interation of server's main loop. 
Seems like I didn't gave it in example. Let me explain.
What i've mean is that server has main fiber with main loop, fiber for handling new users and channel for new sockets, server has fiber per client and channel per client (or one main channel) for messages from clients.

**Comment by asterite** [comment]
Oh, instead of:

```crystal
  def start
    start_new_users_handling

    loop do
      handle_new_users

      # process messages from clients
    end
  end
```

maybe you can do:

```crystal
  def start
    start_new_users_handling

    spawn do
      loop do
        handle_new_users

        # process messages from clients
      end
    end
  end
```

and use `receive` in `handle_new_users`?

**Comment by Odebe** [comment]
Yeah, but what if there is `ClientHandler` that also have loop and recieves messages from other clients (from main servers loop thorugh channel) and messages from fiber created for socket.
I mean `ClientHandler`can't be blocked on receiving messages, it should get messages from server and send them to client even if there no messages from sockets fiber.

Maybe it just my bad architecture desigh decisions.

**Comment by asterite** [comment]
I think this is more like a thing to discuss in the forums, which might be more active than the issue tracker, at least for discussing these things.

The thing is, there's no such thing in Go and I'm pretty sure you can do all kinds of servers in Go, so my guess is that a non-blocking `receive` is not really needed.

**Comment by RX14** [comment]
there's always `select` with `else`...

**Comment by rdp** [comment]
I noticed the other day that the documentation for "Channel" is pretty sparse...if you google "crystal channel" it takes you to this page: https://crystal-lang.org/api/0.31.1/Channel.html which doesn't even explain that you can use select on them, how to make a Nil channel, buffered, etc...might be nice to expand sometime :)
